File: state.py
from random import randint
from random import shuffle
from math import ceil
from helper import isPrime

class State:
    def __init__(self, bin1 = [], bin2 = [], bin3 = [], l = []):
        self.bin1 = list(bin1)
        self.bin2 = list(bin2)
        self.bin3 = list(bin3)
        self.dic = {}
        if (not bin1):
            temp_l = list(l)
            for i in range(0,len(temp_l)):
                if i % 3 == 0:
                    self.bin1.append(int(temp_l[i]))
                elif i % 3 == 1:
                    self.bin2.append(int(temp_l[i]))
                else:
                    self.bin3.append(int(temp_l[i]))
        for i in range(-9, 10):
            self.dic[i] = 0
        for i in self.bin1:
            self.dic[i] += 1
        for i in self.bin2:
            self.dic[i] += 1
        for i in self.bin3:
            self.dic[i] += 1
        self.length = len(self.bin1) * 3
    
    def setscore(self):
        self.sc = self.score()

    def score(self):
        score1 = 0
        score2 = 0
        score3 = 0
        sign = 1
        for num in self.bin1:
            score1 = score1 + num*sign
            sign = sign*-1
        for i in range(0, len(self.bin2) - 1):
            if self.bin2[i+1] > self.bin2[i]:
                score2 = score2 + 3
            elif self.bin2[i+1] == self.bin2[i]:
                score2 = score2 + 5
            else:
                score2 = score2 - 10
        for i in range(0, int(len(self.bin3)/2)):
            if self.bin3[i] < 0:
                score3 = score3 - 2
            elif isPrime(self.bin3[i]):
                score3 = score3 + 4
            else:
                score3 = score3 - self.bin3[i]
        for i in range(int(ceil(len(self.bin3)/2)), len(self.bin3)):
            if self.bin3[i] < 0:
                score3 = score3 + 2
            elif isPrime(self.bin3[i]):
                score3 = score3 - 4
            else:
                score3 = score3 + self.bin3[i]
        return score1 + score2 + score3

    def isLegal(self):
        if (len(self.bin1) == len(self.bin2)) and (len(self.bin1) == len(self.bin3)) and (len(self.bin3) == len(self.bin2)):
            return True
        return False

    def __eq__(self, other):
        if (len(self.bin1) == len(other.bin1)) and (len(self.bin2) == len(other.bin2)) and (len(self.bin3) == len(other.bin3)):
            return other.dic == self.dic
        return False
    
    def toList(self):
        return self.bin1 + self.bin2 + self.bin3
    
    def newState(self):
        one = randint(1,3)
        two = randint(1,3)
        bin1 = list(self.bin1)
        bin2 = list(self.bin2)
        bin3 = list(self.bin3)
        if one == 1:
            binone = bin1
        elif one == 2:
            binone = bin2
        else:
            binone = bin3
        if two == 1:
            bintwo = bin1
        elif two == 2:
            bintwo = bin2
        else:
            bintwo = bin3
        i, j = randint(0, len(bin1) - 2), randint(0, len(bin1) - 1)
        if i == j:
            i+=1
        else:
            i,j = j,i
        temp = binone[i]
        binone[i] = bintwo[j]
        bintwo[j] = temp
        return State(bin1, bin2, bin3)

File: C-version/state.c
#include "state.h"
#include "helper.h"
#include <time.h>
#include <stdlib.h>
#include <string.h>

// fill an empty state's bins and dictionary
void init_state(struct State *state, int binsize, int *list) {
    state->binsize = binsize;
    state->bins = calloc(binsize * 3, sizeof(int));
    for (int i = 0; i < binsize*3; i++) {
        state->bins[i] = list[i];
        state->dic[list[i]+9]++;
    }
    // memcpy(state->bins, list, binsize * 3 * sizeof(int));
}

// remove an empty state's bins
void delete_state(struct State *state) {
    free(state->bins);
}

// shuffles a state in place
// shuffle algorithm from stack overflow
void shuffle(struct State *state) {
    int n = state->binsize * 3;
    for (int i = 0; i < n - 1; i++) {
        int j = i + rand() / (RAND_MAX / (n - i) + 1);
        int t = state->bins[j];
        state->bins[j] = state->bins[i];
        state->bins[i] = t;
    }
}

int score(struct State *state) {
    int score1 = 0;
    int score2 = 0;
    int score3 = 0;
    int sign = 1;
    for (int i = 0; i < state->binsize; i++) {
        score1 += sign*state->bins[i];
        sign *= -1;
    }
    for (int i = state->binsize; i < state->binsize*2 - 1; i++) {
        if (state->bins[i] < state->bins[i+1]) {
            score2 += 3;
        }
        else if (state->bins[i] == state->bins[i+1]) {
            score2 += 5;
        }
        else {
            score2 -= 10;
        }
    }
    for (int i = state->binsize*2; i < state->binsize*2 + state->binsize / 2; i++) {
        if (state->bins[i] < 0) {
            score3 -= 2;
        }
        else if (is_prime(state->bins[i])) {
            score3 += 4;
        }
        else {
            score3 -= state->bins[i];
        }
    }
    for (int i = state->binsize*2 + 1 + (state->binsize-1)/2; i < state->binsize*3; i++) {
        if (state->bins[i] < 0) {
            score3 += 2;
        }
        else if (is_prime(state->bins[i])) {
            score3 -= 4;
        }
        else {
            score3 += state->bins[i];
        }
    }
    state->score = score1+score2+score3;
    return score1 + score2 + score3;

}

// fill new_state with state, optional random swap
void new_state(struct State *new_state, struct State *state, int swap) {
    int *new_list = calloc(state->binsize*3, sizeof(int));
    memcpy(new_list, state->bins, state->binsize*3 * sizeof(int));
    if (swap) {
        int i, j;
        do {
            i = rand() % (state->binsize*3);
            j = rand() % (state->binsize*3);
        } while (i == j);
        int temp = new_list[i];
        new_list[i] = new_list[j];
        new_list[j] = temp;
    }
    init_state(new_state, state->binsize, new_list);
    free(new_list);
}

